Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -49,87 +55,246 @@ lemma primesBelow_succ (n : ℕ) :
 lemma not_mem_primesBelow (n : ℕ) : n ∉ primesBelow n :=
   fun hn ↦ (lt_of_mem_primesBelow hn).false
 
+
+/-!
+### `s`-factored numbers
+-/
+
+/-- `factoredNumbers s`, for a finite set `s` of natural numbers, is the set of positive natural
+numbers all of whose prime factors are in `s`. -/
+def factoredNumbers (s : Finset ℕ) : Set ℕ := {m | m ≠ 0 ∧ ∀ p ∈ factors m, p ∈ s}
+
+lemma mem_factoredNumbers {s : Finset ℕ} {m : ℕ} :
+    m ∈ factoredNumbers s ↔ m ≠ 0 ∧ ∀ p ∈ factors m, p ∈ s :=
+  Iff.rfl
+
+/-- Membership in `Nat.factoredNumbers n` is decidable. -/
+instance (s : Finset ℕ) : DecidablePred (· ∈ factoredNumbers s) :=
+  inferInstanceAs <| DecidablePred fun x ↦ x ∈ {m | m ≠ 0 ∧ ∀ p ∈ factors m, p ∈ s}
+
+/-- A number that divides an `s`-factored number is itself `s`-factored. -/
+lemma mem_factoredNumbers_of_dvd {s : Finset ℕ} {m k : ℕ} (h : m ∈ factoredNumbers s)
+    (h' : k ∣ m) (hk : k ≠ 0) :
+    k ∈ factoredNumbers s := by
+  rw [mem_factoredNumbers] at h ⊢
+  obtain ⟨h₁, h₂⟩ := h
+  refine ⟨hk, fun p hp ↦ h₂ p ?_⟩
+  rw [mem_factors <| by assumption] at hp ⊢
+  exact ⟨hp.1, hp.2.trans h'⟩
+
+/-- `m` is `s`-factored if and only if `m` is nonzero and all prime divisors `≤ m` of `m`
+are in `s`. -/
+lemma mem_factoredNumbers_iff_forall_le {s : Finset ℕ} {m : ℕ} :
+    m ∈ factoredNumbers s ↔ m ≠ 0 ∧ ∀ p ≤ m, p.Prime → p ∣ m → p ∈ s := by
+  simp_rw [mem_factoredNumbers, mem_factors']
+  exact ⟨fun ⟨H₀, H₁⟩ ↦ ⟨H₀, fun p _ hp₂ hp₃ ↦ H₁ p ⟨hp₂, hp₃, H₀⟩⟩,
+    fun ⟨H₀, H₁⟩ ↦
+      ⟨H₀, fun p ⟨hp₁, hp₂, hp₃⟩ ↦ H₁ p (le_of_dvd (Nat.pos_of_ne_zero hp₃) hp₂) hp₁ hp₂⟩⟩
+
+/-- `m` is `s`-factored if and only if all prime divisors of `m` are in `s`. -/
+lemma mem_factoredNumbers' {s : Finset ℕ} {m : ℕ} :
+    m ∈ factoredNumbers s ↔ ∀ p, p.Prime → p ∣ m → p ∈ s := by
+  obtain ⟨p, hp₁, hp₂⟩ := exists_infinite_primes (1 + Finset.sup s id)
+  rw [mem_factoredNumbers_iff_forall_le]
+  refine ⟨fun ⟨H₀, H₁⟩ ↦ fun p hp₁ hp₂ ↦ H₁ p (Nat.le_of_dvd (Nat.pos_of_ne_zero H₀) hp₂) hp₁ hp₂,
+         fun H ↦ ⟨fun h ↦ ?_, fun p _ ↦ H p⟩⟩
+  exact ((Finset.le_sup (f := @id ℕ) <| (H p hp₂ <| h.symm ▸ dvd_zero p)).trans_lt (lt_one_add _)
+            |>.trans_le hp₁).false
```

Comment:
This feels like it's golfed a bit much. Maybe put `lt_irrefl` in the `refine` above and use `calc` here?

Reply:
It's slightly slower, but I agree that it is clearer.